File: src/main/java/net/snipsniper/secrets/games/BGamePiece.java
package net.snipsniper.secrets.games;

import net.snipsniper.utils.ImageUtils;

import java.awt.*;
import java.awt.image.BufferedImage;

public class BGamePiece {
    private final BGame game;

    public int[][] figure;
    private int posX;
    private int posY = 1;

    private int rotationCooldown = 0;
    private final int rotationCooldownMax = 10;

    private int moveCooldown = 0;
    private final int moveCooldownMax = 10;

    public final int index;

    public BGamePiece(BGame game) {
        index = BGame.randomRange(0,game.getResources().getSize() - 1);
        //index = 0;
        figure = game.getResources().getPiece(index);
        this.game = game;

        posX = (this.game.BOARD_WIDTH/2) - figure[0].length/2;

        for(int y = 0; y < figure.length; y++) {
            for(int x = 0; x < figure[y].length; x++) {
                if(game.getBoard()[posX + x][posY + y] != null) {
                    game.gameOver();
                }
            }
        }
        
        //Important:
        //When looping through int[][] figure you need to do figure[y][x]
    }

    public boolean update() {
        if(rotationCooldown > 0)
            rotationCooldown--;
        if(moveCooldown > 0)
            moveCooldown--;

        return checkCollision();
    }

    public boolean moveDown() {
        if(!checkCollision()) {
            posY++;
            return false;
        }
        return true;
    }

    public BufferedImage getRawImage(int size) {
        BufferedImage image = new BufferedImage(figure[0].length * size, figure[1].length * size, BufferedImage.TYPE_INT_ARGB);
        Graphics g = image.createGraphics();
        for(int y = 0; y < figure[0].length; y++) {
            for(int x = 0; x < figure.length; x++) {
                if(figure[y][x] != 0) {
                    g.drawImage(game.getResources().getImage(index), x * size, y * size, size, size, null);
                }
            }
        }
        g.dispose();
        return ImageUtils.Companion.trimImage(image);
    }

    //returns true if the piece hit something
    public boolean checkCollision() {
        for(int y = 0; y < figure[0].length; y++) {
            for(int x = 0; x < figure.length; x++) {
                if(figure[y][x] != 0) {
                    final int nPosX = posX + x;
                    final int nPosY = posY + y;
                    if(!(nPosX <= -1) && !(nPosX >= game.BOARD_WIDTH) && !(nPosY <= 0) && !(nPosY >= game.BOARD_HEIGHT - 1)) {
                        BGameBlock cBlock = game.getBoard()[posX + x][posY + y + 1];
                        if(cBlock != null)
                            return true;
                    }
                    if(posY + y >= game.BOARD_HEIGHT-1)
                        return true;
                }
            }
        }
        return false;
    }

    public void hit() {
        for(int y = 0; y < figure[0].length; y++) {
            for(int x = 0; x < figure.length; x++) {
                if(figure[y][x] != 0) {
                    final int nPosX = posX + x;
                    final int nPosY = posY + y;
                    if(!(nPosX < 0 || nPosX >= game.BOARD_WIDTH || nPosY < 0 || nPosY >= game.BOARD_HEIGHT))
                        game.getBoard()[posX + x][posY + y] = new BGameBlock(index);
                }
            }
        }
    }

    private boolean checkMoveCollision(int dir) {
        for(int y = 0; y < figure[0].length; y++) {
            for(int x = 0; x < figure.length; x++) {
                final int piecePosX = posX + x + dir;
                final int piecePosY = posY + y;

                if(figure[y][x] != 0) {
                    if(piecePosX <= -1)
                        return false;
                    if(piecePosX >= game.BOARD_WIDTH)
                        return false;
                    if(game.getBoard()[piecePosX][piecePosY] != null)
                        return false;
                }
            }
        }
        return true;
    }

    public void move(int dir) {
        if(dir == 1 && moveCooldown == 0) {
            if(checkMoveCollision(dir)) {
                posX++;
                moveCooldown = moveCooldownMax;
            }
        } else if (dir == -1 && moveCooldown == 0) {
            if(checkMoveCollision(dir)) {
                posX--;
                moveCooldown = moveCooldownMax;
            }
        }
    }

    private boolean checkRotation(int[][] newFigure) {
        for(int y = 0; y < newFigure[0].length; y++) {
            for(int x = 0; x < newFigure.length; x++) {
                final int nPosX = posX + x;
                final int nPosY = posY + y;
                if((nPosX <= -1) || (nPosX >= game.BOARD_WIDTH) || (nPosY <= 0) || (nPosY >= game.BOARD_HEIGHT - 1)) {
                    if(newFigure[y][x] == 1)
                        return false;
                }
                if(!(nPosX < 0 || nPosX >= game.BOARD_WIDTH || nPosY < 0 || nPosY >= game.BOARD_HEIGHT))
                    if(game.getBoard()[nPosX][nPosY] != null)
                        return false;
            }
        }
        return true;
    }

    public void rotate(int dir) {
        if(rotationCooldown == 0) {
            int[][] newFigure = rotateMatrix(figure, dir);
            if(checkRotation(newFigure)) {
                figure = newFigure;
                rotationCooldown = rotationCooldownMax;
            }
        }

    }

    private static int[][] rotateMatrix(int[][] figure2, int dir) {
        int[][] rotated = new int[figure2[0].length][figure2.length];
        for (int i = 0; i < figure2[0].length; ++i) {
            for (int j = 0; j < figure2.length; ++j) {
                if(dir == 1)
                    rotated[i][j] = figure2[figure2.length - j - 1][i];
                else if(dir == -1)
                    rotated[i][j] = figure2[j][figure2[0].length - i - 1];
            }
        }
        return rotated;
    }

    public int getPosX() {
        return posX;
    }

    public int getPosY() {
        return posY;
    }
}

File: src/main/kotlin/net/snipsniper/secrets/games/BGamePiece.kt
package net.snipsniper.secrets.games

import net.snipsniper.utils.ImageUtils
import net.snipsniper.utils.trim
import org.capturecoop.ccutils.math.CCVector2Int
import java.awt.image.BufferedImage

class BGamePiece(private val game: BGame) {
    var figure: Array<IntArray>
    var posX: Int
        private set
    var posY = 1
        private set
    private var rotationCooldown = 0
    private val rotationCooldownMax = 10
    private var moveCooldown = 0
    private val moveCooldownMax = 10
    val index: Int = BGame.randomRange(0, game.resources!!.getSize() - 1)

    init {
        figure = game.resources!!.getPiece(index)
        posX = BGame.BOARD_WIDTH / 2 - figure[0].size / 2
        for (y in figure.indices) {
            for (x in figure[y].indices) {
                if (game.board[posX + x][posY + y] != null) {
                    game.gameOver()
                }
            }
        }

        //Important:
        //When looping through int[][] figure you need to do figure[y][x]
    }

    fun update(): Boolean {
        if (rotationCooldown > 0) rotationCooldown--
        if (moveCooldown > 0) moveCooldown--
        return checkCollision()
    }

    fun moveDown() = when(checkCollision()) {
        false -> {
            posY++
            false
        }
        true -> true
    }

    fun getRawImage(size: Int): BufferedImage {
        return ImageUtils.newBufferedImage(figure[0].size * size, figure[1].size * size) {
            for (y in figure[0].indices) {
                for (x in figure.indices) {
                    if(figure[y][x] == 0) continue
                    it.drawImage(game.resources!!.getImage(index), x * size, y * size, size, size, null)
                }
            }
        }.trim()
    }

    //returns true if the piece hit something
    fun checkCollision(): Boolean {
        for (y in figure[0].indices) {
            for (x in figure.indices) {
                if(figure[y][x] == 0) continue
                CCVector2Int(posX + x, posY + y).also { v ->
                    if (
                        v.x > -1 && v.x < BGame.BOARD_WIDTH &&
                        v.y > 0 && v.y < BGame.BOARD_HEIGHT - 1 &&
                        game.board[posX + x][posY + y + 1] != null
                    ) return true
                    if (posY + y >= BGame.BOARD_HEIGHT - 1) return true
                }
            }
        }
        return false
    }

    fun hit() {
        for (y in figure[0].indices) {
            for (x in figure.indices) {
                if(figure[y][x] == 0) continue
                CCVector2Int(posX + x, posY + y).also { v ->
                    if(!(v.x < 0 || v.x >= BGame.BOARD_WIDTH || v.y < 0 || v.y >= BGame.BOARD_HEIGHT))
                        game.board[v.x][v.y] = BGameBlock(index)
                }
            }
        }
    }

    private fun checkMoveCollision(dir: Int): Boolean {
        for (y in figure[0].indices) {
            for (x in figure.indices) {
                if(figure[y][x] == 0) continue
                CCVector2Int(posX + x + dir, posY + y).also { v ->
                    if (v.x <= -1) return false
                    if (v.x >= BGame.BOARD_WIDTH) return false
                    if (game.board[v.x][v.y] != null) return false
                }
            }
        }
        return true
    }

    fun move(dir: Int) {
        if(moveCooldown != 0 || !checkMoveCollision(dir)) return
        when(dir) {
            1 -> {
                posX++
                moveCooldown = moveCooldownMax
            }
            -1 -> {
                posX--
                moveCooldown = moveCooldownMax
            }
        }
    }

    private fun checkRotation(newFigure: Array<IntArray>): Boolean {
        for (y in newFigure[0].indices) {
            for (x in newFigure.indices) {
                CCVector2Int(posX + x, posY + y).also { v ->
                    if (v.x <= -1 || v.x >= BGame.BOARD_WIDTH || v.y <= 0 || v.y >= BGame.BOARD_HEIGHT - 1) {
                        if (newFigure[y][x] == 1) return false
                    }
                    if(!(v.x < 0 || v.x >= BGame.BOARD_WIDTH || v.y < 0 || v.y >= BGame.BOARD_HEIGHT))
                        if(game.board[v.x][v.y] != null)
                            return false
                }
            }
        }
        return true
    }

    fun rotate(dir: Int) {
        if(rotationCooldown != 0) return
        rotateMatrix(figure, dir).also {
            if (checkRotation(it)) {
                figure = it
                rotationCooldown = rotationCooldownMax
            }
        }
    }

    companion object {
        private fun rotateMatrix(figure2: Array<IntArray>, dir: Int): Array<IntArray> {
            val rotated = Array(figure2[0].size) { IntArray(figure2.size) }
            for (i in figure2[0].indices) {
                for (j in figure2.indices) {
                    if (dir == 1) rotated[i][j] = figure2[figure2.size - j - 1][i] else if (dir == -1) rotated[i][j] =
                        figure2[j][figure2[0].size - i - 1]
                }
            }
            return rotated
        }
    }
}